### DISCUSSION TITLE
[Feature] Generators Local Aliases / Generator Priorities

### DISCUSSION THREAD

**Discussion by Char2sGu** [nrwl/nx#22934] [Feature Requests]
I am migrating from Angular CLI to NX. Things are mostly satisfying, except for being required to write the verbose `collection:generator`  syntax for almost every code generations.

Scaffolding components, directives, and pipes are the most frequent operations within my workflow. With Angular CLI, I was able to write `ng --project foo g c bar` for a generation, while with Nx I have to use `nx --project foo g @schematics/angular:c bar` for the same operation.

I was wondering if it would be a good idea to allow the developer to define local aliases for Nx Generators / Angular Schematics in `nx.json`. The syntax might be something like:

nx.json
```json
"generatorAliases": {
    "c": "@schematics/angular:component",
    "d": "@schematics/angular:directive"
}
```

By doing so, `nx g c` is equivalent to `nx g @schematics/angular:component`.

---

Alternatively, allow the developer to specify _plugin priorities_:

nx.json
```json
"pluginPriorities": ["@schematics/angular"]
```

By doing so, for any generator names that do not have a collection explicitly specified, Nx will go through the list of plugins specified in `pluginPriorities` and use the first generator that matches.

**Comment by jaysoo** [comment]
The Nx generators do have aliases defined, for example `c` for `@nx/angular:component`. https://github.com/nrwl/nx/blob/master/packages/angular/generators.json#L22

So `nx g c` would run `@nx/angular:component` as long as `c` isn't matched to another generator. If there are multiple aliased generators, than Nx will prompt you to pick on.

**Reply by Char2sGu** [reply]
> as long as c isn't matched to another generator

The thing is one alias is frequently matched by multiple generators, and that's why we need _local alias_ for a better DX.

Example: I have both `@schematics/angular` and `@nx/angular` installed, but I always prefer the schematics from `@schematics/angular` over the generators from `@nx/angular`. By `nx g c`, I always mean `nx g @schematics/angular:c`, but since the alias `c` is registered by both `@schematics/angular:component` and `@nx/angular:component`, I have to either type the full path or manually select the collect in the prompt, which is frustrating.

Another Example: I have both `@nx/js` and `@nx/angular` installed in an Angular workspace. By `nx g lib`, I always mean to generate an Angular library with `nx g @nx/angular:library`, but since the alias `lib` refer to both `@nx/js:library` and `@nx/angular:library`, I have to take extra actions/typings.

**Reply by Char2sGu** [reply]
As discussed in #22625, it is *necessary* to install `@nx/angular` in order to execute schematics, thus the alias conflicts are inevitable.

**Reply by jaysoo** [reply]
Got it. Is this a dev ergonomics concern? Or is it important that only one generator is chosen from a particular set?

The assumption from us is that people know which one from the prompt to choose, and it's better than not prompting, in case they do want to choose different ones.

**Reply by Char2sGu** [reply]
It is a dev ergonomics concern. If the developer is sure that he only needs the alias "c" to refer to `@schematics/angular:component`, prompting a selection dialog would be extraneous and decrease productivity.

I understand your assumption, while this "local alias" thing is defined by the developer, and when not specified, Nx should still prompt the selection dialog to ask for further details.

**Reply by Char2sGu** [reply]
Hey @jaysoo can I expect any updates on this topic?
